plutus-scb/src/Cardano/Protocol/Socket/Server.hs

```diff
@@ -32,173 +41,392 @@ import           Ouroboros.Network.Protocol.Handshake.Version
+                -- state to remain synchronised.
+                writeTChan blocks block
+                writeTQueue commandQueue (Right (BlockAdded block))
+        -- A client sent a response instead of a request.
+        Right response ->
+            atomically $ writeTQueue commandQueue (Right (IgnoredResponse response))
+
+{- | Start the server in a new thread, and return a server handler
+     used to control the server -}
+runServerNode ::
+    FilePath
+ -> ChainState
+ -> IO ServerHandler
+runServerNode socketPath initialState = do
+    serverState  <- initialiseInternalState initialState
+    commandQueue <- newTQueueIO
+    _ <- forkIO . void    $ protocolLoop  socketPath   serverState
+    _ <- forkIO . forever $ handleCommand commandQueue serverState
+    pure ServerHandler {
+        shSocketPath   = socketPath,
+        shCommandQueue = commandQueue
+    }
+
+-- * Internal state
+
+{- Store a channel (that is *never* cleared, for now) containing all
+   the blocks, and the state required for validating the next block. -}
+data InternalState = InternalState {
+    isBlocks :: TChan Block,
+    isState  :: MVar ChainState
+}
+
+initialiseInternalState ::
+    ChainState
+ -> IO InternalState
+initialiseInternalState chainState = do
+    blocks  <- newTChanIO
+    state   <- newMVar chainState
+    _ <- traverse (atomically . writeTChan blocks)
+             (view Chain.chainNewestFirst chainState)
+    pure InternalState {
+        isBlocks = blocks,
+        isState  = state
+    }
+
+-- * ChainSync protocol
+
+{- A monad for running all code executed when a state
+   transition is invoked. It makes the implementation of
+   state transitions easier to read. -}
+
+type ChainSyncMonad = ReaderT InternalState IO
+
+runChainSync :: InternalState -> ChainSyncMonad a -> IO a
+runChainSync = flip runReaderT
+
+{- The initial state of the protocol. You can move into
+   requesting the next block or reset state by searching for an
+   intersection. -}
+idleState ::
+    ( MonadReader InternalState m
+    , MonadIO m )
+ => Maybe LocalChannel
+ -> m (ServerStIdle Block Tip m ())
+idleState (Just channel) =
+    pure ServerStIdle {
+        recvMsgRequestNext = nextState channel,
+        recvMsgFindIntersect = findIntersect,
+        recvMsgDoneClient = return ()
+    }
+idleState Nothing = undefined
+
+{- Get the next block, either immediately (the Just/Left branch)
+   or within a monad (IO, in our case) where you can wait for the
+   next block (Nothing/Right branch) -}
+nextState ::
+    ( MonadReader InternalState m
+    , MonadIO m )
+ => LocalChannel
+ -> m (Either (ServerStNext Block Tip m ())
+              (m (ServerStNext Block Tip m ())))
+nextState localChannel@(LocalChannel channel) = do
+    chainState <- ask <&> isState
+    tip <- head . view Chain.chainNewestFirst <$>
+           liftIO (readMVar chainState)
+    (liftIO . atomically $ tryReadTChan channel) >>= \case
```

Comment:
```suggestion
    maybeBlock <- liftIO . atomically $ tryReadTChan channel
    case maybeBlock of
```

Reply:
Is there a reason why you don't like this usage of `LambdaCase`?

Reply:
It's not that I don't like, it's `>>=` with complex expressions on either side. I just find it usually hard to read.